components/calendar/src/dangi.rs

```diff
@@ -0,0 +1,1238 @@
+        ArithmeticDate::new_from_lunar_ordinals(year, month, day)
+            .map(ChineseBasedDateInner)
+            .map(DangiDateInner)
+            .map(|inner| Date::from_raw(inner, Dangi))
+    }
+}
+
+impl DateTime<Dangi> {
+    /// Construct a new Dangi DateTime from integers. See [`try_new_dangi_date`].
+    ///
+    /// ```rust
+    /// use icu::calendar::DateTime;
+    ///
+    /// let dangi_datetime = DateTime::try_new_dangi_datetime(4356, 6, 6, 13, 1, 0)
+    ///     .expect("Failed to initialize Dangi DateTime instance.");
+    ///
+    /// assert_eq!(dangi_datetime.date.year().number, 4356);
+    /// assert_eq!(dangi_datetime.date.year().related_iso, Some(2023));
+    /// assert_eq!(dangi_datetime.date.year().cyclic, Some(40));
+    /// assert_eq!(dangi_datetime.date.month().ordinal, 6);
+    /// assert_eq!(dangi_datetime.date.day_of_month().0, 6);
+    /// assert_eq!(dangi_datetime.time.hour.number(), 13);
+    /// assert_eq!(dangi_datetime.time.minute.number(), 1);
+    /// assert_eq!(dangi_datetime.time.second.number(), 0);
+    /// ```
+    pub fn try_new_dangi_datetime(
+        year: i32,
+        month: u8,
+        day: u8,
+        hour: u8,
+        minute: u8,
+        second: u8,
+    ) -> Result<DateTime<Dangi>, CalendarError> {
+        Ok(DateTime {
+            date: Date::try_new_dangi_date(year, month, day)?,
+            time: types::Time::try_new(hour, minute, second, 0)?,
+        })
+    }
+}
+
+impl ChineseBased for Dangi {
+    fn location(fixed: RataDie) -> Location {
+        if fixed < FIXED_1908 {
+            KOREAN_LOCATION_ORIGINAL
+        } else if fixed < FIXED_1912 {
+            KOREAN_LOCATION_1908
+        } else if fixed < FIXED_1954 {
+            KOREAN_LOCATION_1912
+        } else if fixed < FIXED_1961 {
+            KOREAN_LOCATION_1954
+        } else {
+            KOREAN_LOCATION_1961
+        }
+    }
+
+    const EPOCH: RataDie = KOREAN_EPOCH;
+
+    // Unwrap can be used for this function because it is only called internally when
+    // generating dates from a valid year, month, and day
+    #[allow(clippy::unwrap_used)]
+    fn new_chinese_based_date(year: i32, month: u8, day: u8) -> ChineseBasedDateInner<Dangi> {
+        ChineseBasedDateInner(ArithmeticDate::new_unchecked(year, month, day))
+    }
+}
+
+impl Dangi {
+    /// Get the ISO Date of Seollal (Korean Lunar New Year) on or before the given ISO Date `iso`.
+    ///
+    /// ```rust
+    /// use icu::calendar::Date;
+    /// use icu::calendar::dangi::Dangi;
+    ///
+    /// let iso = Date::try_new_iso_date(2023, 6, 18)
+    ///     .expect("Failed to initialize ISO Date instance.");
+    /// let seollal = Dangi::seollal_on_or_before_iso(iso);
+    ///
+    /// assert_eq!(seollal.year().number, 2023);
+    /// assert_eq!(seollal.month().ordinal, 1);
+    /// assert_eq!(seollal.day_of_month().0, 22);
+    /// ```
+    pub fn seollal_on_or_before_iso(iso: Date<Iso>) -> Date<Iso> {
```

Comment:
Question: do we want this as a pub fn?

Reply:
I made it pub since the equivalent in Chinese `chinese_new_year_on_or_before_iso` is also pub.